### DISCUSSION TITLE
Sensitivities of alphaPorosity are messed up on mpirun

### DISCUSSION THREAD

**Discussion by schewef** [mdolab/dafoam#102] [Bug] [answered]
## Description
During sensitivity calculation of alphaPorosity in the pitzDaily tutorial case the sensitivity seem to be messed up if an mpirun in performed.

The first picture below shows an sensitivity calculation executed on 4 cores with:
mpirun -np 4 python runScript.py 2>&1 | tee logOpt.txt

![image](https://user-images.githubusercontent.com/43476213/112177199-0449e000-8bf9-11eb-861b-ecc8c84184d5.png)

The second picture shows an execution on one core with:
python runScript.py 2>&1 | tee logOpt.txt

![image](https://user-images.githubusercontent.com/43476213/112177373-23487200-8bf9-11eb-96ce-32ca359f9a1f.png)

The sensitivities in the first case look unphysical. Maybe it's an domain decomposition issue.

### Steps to reproduce issue

Attached you can find two models "mpirun" and "no_mpirun". The only difference is the argument the function was called with.
The sensitivities are written to the "0" folder in OpenFOAM format for easiely importing them into e.g. ParaView.

For reproduction do the following
1. Delete the "sensitivity" file in "0" folder
2. Perform "mpirun -np 4 python runScript.py 2>&1 | tee logOpt.txt" in "mpirun"
3. Perform "python runScript.py 2>&1 | tee logOpt.txt" in "no_mpirun"
4. Compare the sensitiviy results stored in "0" after execution

Files for reproduction:

[BugReport.zip](https://github.com/mdolab/dafoam/files/6191140/BugReport.zip)

**Comment by friedenhe** [comment, marked as answer]
@schewef I moved this from "Issues" to "Discussion" because I think the workaround may be helpful for others.
 
We checked the sensitivity between the serial and parallel case and they matched. It seems that you output sens["PL"]["alphaPorosity"] for the parallel and serial cases to 0/sensitivity. However, the order of the alpha field is different in parallel and in serial. In order words, if you check the order of the 0/alphaPorosity variable, it is not simply adding processor0/alphaPorosity, processor1/alphaPorosity, processor2/alphaPorosity, processor3/alphaPorosity.

Right now, there is no easy way to visualize the sensitivity field in parallel. The workaround is: 

- Download the latest version of DAFoam on Github.
- Run a case with `--task=runAdjoint`. You will see a file totalDerivHist.txt (in JSON format) that saves the totalDeriv (sensitivity). Remove the header "Optimization Iteration 001" 
- Modify the runScript.py to load this totalDerivHist information in and assign it to the initial field of alphaPorosity, something like:

<pre>
with open("dvParallel/totalDerivHist.txt") as f:
    alphaInit = json.load(f)
alphaPorosity0 = alphaInit["PL"]["alphaPorosity"]

DVGeo.addGeoDVGlobal("alphaPorosity", value=alphaPorosity0, func=alphaPorosity, lower=0.0, upper=1000.0, scale=0.001)
</pre>

- Then, run the job with `--task=runPrimal`. The totalDeriv will be saved as the alphaPorosity variable and save to the disk when the flow simulation is done (you can just run for one step). You will be able to view it in Paraview. Note that alphaPorosity will be written to the processor* folders for the parallel cases.

If you do the above, the serial and parallel cases should match. Let me know if it works.

**Reply by schewef** [reply]
@friedenhe Thank you very much for the reply, I really appreciate that! Just tried your workaround. It works perfectly!

**Reply by friedenhe** [reply]
@schewef FYI, the latest version of DAFoam can write the surface and volume sensitivity to files during optimization, check [this](https://dafoam.github.io/mydoc_get_started_faq.html#how-to-get-sensitivity-maps).

**Reply by schewef** [reply]
Great! Thanks for the information!

**Reply by schewef** [reply]
@friedenhe Is there an simple way to get the mapping between the index of each cell in the design variable / sensitivity vector and the index of the cell in the OpenFOAM mesh in parallel cases?

Or can I write e.g. the alphaPorosity definded by "DVGeo.setDesignVars(xDV)" to the "process" folders without running the primal solution?

I try to implement e.g. a volume fraction constraint for topology optimization and struggle with the fact that the cell vector extracted from the OpenFOAM mesh is different from the one used in DAFoam in parallel cases. Therefore I am searching for a methodology to map which cell in the OpenFOAM mesh belongs to which porosity / sensitivity in the DAFoam vector.

**Reply by schewef** [reply]
I think I just found the solution by myself. The "setFieldValue4GlobalCellI" is made for handling this issure if I get the discription right.

**Reply by qwefhajk** [reply]
@schewef Do you master the function "setFieldValue4GlobalCellI"? I do not know how it makes a difference in this framework

**Reply by schewef** [reply]
@qwefhajk The function can be used to assign a vector of e.g. alphaPorosity values from the "global" cell index defined in the OpenFOAM mesh to the corresponding cell index of the vector DAFoam uses. 

Therefore it is possible to implement a methodology to extract a mapping from OpenFOAM cell index to DAFoam cell index and to reshape your sensitivity vector according to the mapping.

**Reply by qwefhajk** [reply]
Thanks for your reply! So What is the cell index in DAFoam that we should follow. I mean, the input of setFieldValue4GlobalCellI   seems ordinary and it dose not disclose the information on indexing.

**Reply by schewef** [reply]
Thats true, the function itself does not disclose the information on indexing, but you can use it to implement a methodology to get it.

You can simply use the "setFieldValue4GlobalCellI" function to assign a vector in "global shape" and compare it with the vector dafoam exports after decomposition.
Afterwards you know which "global index" belongs to which "dafoam index".

To be honest the implementation is a little tricky and the solution is not very elegant but it works in the end. For sure there are better solutions but I did not invest the time to go deeper inside the code to implement one.

**Reply by qwefhajk** [reply]
Thanks a lot!! I will try out your suggestion!
